File: lab1/zad2.py
import numpy as np
import matplotlib.pyplot as plt

fs2 = 200
fs1 = 10000

dt2 = 1/fs2
dt1 = 1/fs1

t_fs2 = np.arange(0, 0.1, dt2)
t_fs1 = np.arange(0, 0.1, dt1)

y_fs2 = 230 * np.sin(2 * np.pi * 50 * t_fs2)
y_fs1 = 230 * np.sin(2 * np.pi * 50 * t_fs1)

y_out = np.zeros(len(t_fs1))

for i in range(len(t_fs1)):
    val = 0
    for n in range(len(t_fs2)):
        T1 = 1/fs2
        T2 = 1/fs1
        t = i * T2
        nT = n * T1
        y = np.pi/T1 * (t - nT)
        sinc = 1
        if y != 0:
            sinc = np.sin(y)/y
        val = val + y_fs2[n] * sinc
    y_out[i] = val

plt.figure()
plt.grid(True)
plt.plot(t_fs1, y_fs1, 'r-')
plt.plot(t_fs1, y_out, 'b-')
plt.xlabel('Czas (s)')
plt.ylabel('Amplituda (V)')
plt.title('Sygnał oryginalny i odtworzony')
plt.legend(['Sygnał oryginalny', 'Sygnał odtworzony'])

errors = np.abs(y_fs1 - y_out)
plt.figure()
plt.grid(True)
plt.plot(t_fs1, errors, 'g-')
plt.xlabel('Czas [s]')
plt.ylabel('Amplituda [V]')
plt.title('Błędy rekonstrukcji sygnału')
plt.show()

File: lab1/zad2.m
clear all; close all; clc;

t = 2*pi;
base_freq = 10;
base_period = 1/base_freq;
base_time = 0:base_period:t;
base_sig = sin(base_time);

rec_freq = base_freq * 50;
rec_period = 1/rec_freq;
rec_time = 0:rec_period:t;
rec_sig = zeros(length(rec_time), 1);

for t_interval = 1:length(rec_time)
    sum = 0;
    for sample = 1:length(base_time)
        y = pi / base_period * ((t_interval-1)*rec_period - (sample-1)*base_period);
        sinc = 1;
        if y ~= 0
            sinc = sin(y) / y;
        end
        sum = sum + base_sig(sample) * sinc;
    end
    rec_sig(t_interval) = sum;
end

plot(base_time, base_sig, 'r-', rec_time, rec_sig, 'b-')
legend('Sygnał oryginalny', 'Sygnał odtworzony')
grid on;
title('Porównanie sygnału oryginalnego i odtworzonego')